lib/smart_proxy_pulp_plugin/pulpcore_plugin.rb

```diff
@@ -6,21 +6,24 @@ class PulpcorePlugin < ::Proxy::Plugin
     default_settings :pulp_url => 'https://localhost',
                      :content_app_url => 'https://localhost:24816/',
                      :mirror => false,
-                     :client_authentication => ['password', 'client_certificate']
+                     :client_authentication => ['password', 'client_certificate'],
+                     :rhsm_url => 'https://localhost/rhsm'
```

Comment:
I question a default value here. For Pulp it's already debatable, but this implies there's always a RHSM URL. If you don't specify it, there's a useless value reported to Katello which can actually break things.

I'd argue for not having a default.

Then we can debate whether it should fail to initialize if no RHSM URL is specified. If it's a required value, you can still keep the validation. If not, we can talk about how Katello should interpret that.

Reply:
> But let's approach it from a different angle: I suspect there are 0 users where `https://localhost/rhsm` is a valid value. If nobody uses a default, the default is wrong. That's why I'd argue for not having a default.

OK, agreed.

> I've asked these exact questions, but haven't really gotten an answer. From a code perspective, you have a content_source and today it assumes it has both. On the other hand, you don't need Candlepin for all content types. For example, with Python content (I think that's being worked on) Candlepin doesn't make sense. Nor for plain files.

So in Katello itself, a repository requires an associated product to be created, and creating the product requires Candlepin.

BUT, smart_proxy itself doesn't have to care about these things, and when I dug into it further there does exist an option to publish repositories of certain types over HTTP.

So in theory you could create repositories that are published over HTTP and sync those to a smart_proxy and that smart_proxy could mirror the repositories.

What is less clear to me is how this would actually work with content views and syncing. This "smart_proxy without rhsm" wouldn't have any clients registering through it, let alone receiving subscriptions, so you would want to avoid syncing any lifecycle environments associated with content view versions containing any content published over HTTPS. It wouldn't break anything, but the content itself would be useless.

Then even if you created separate Lifecycle Environment Paths for the HTTP-only content, it's not clear to me (and I haven't had time to test) what is the behavior of HTTP-published repositories within a content view. I believe, but I am not certain, that anything within a content view is going to be HTTPS only. If that's the case, then even associating the Library environment with the smart_proxy -- which is not recommended because it synchronizes everything under the sun -- would not result in HTTP published repositories on the external smart_proxy.

If I am correct about that, then HTTP publishing is only a feature on the primary smart_proxy associated with Katello itself, and that means HTTP-only repositories is not an use case for smart_proxy_pulp without RHSM, because creating the repository at all on Katello requires a custom product be created first, which requires Candlepin.

Reply:
There are good thought provoking questions throughout this thread but this is not the best place to try and hash out and we should aim for how things definitely work today. To that end, I am in agreement that we should enforce needing an `rhsm_url` and relax that later if any of these other ways of doing things become an option.

> * Is there ever a situation where an user might want to deploy Katello with Pulp but without Candlepin?

If you ask users, there is because they see Candlepin as heavy and unnecessary. Is Katello architected today to allow this and/or working towards this goal? No. So yes, there is a use case for Pulp without Candlepin for non-yum content types but there is no roadmap for it and we should code this accordingly.

> * Is there ever a situation where an user might want to deploy one or more SmartProxies with Pulp but no RHSM support?

Same as above, a user could want to deploy content mirroring and skip using all of the power of activation keys, rhsm registration, certificate handling etc. They would need to then handle a lot out of band, setting up their clients to point at the repositories, deploying without cert guard. In theory, a user would even configure subscription-manager to talk to one smart-proxy for `/rhsm` and get content from another smart-proxy. The Katello workflows do not support this and until we have a first-class solution I think we should not try to code or optimize for them.

Let's stick with the way things work, the rules that bind us today and accept that Pulp and `/rhsm` will be co-located for now. We can always change code down the road if our designs change.

Reply:
Technically what this change allows us to do is to point to the central Candlepin installation if we don't install the reverse proxy so they don't **have** to be co-located. So in that sense, this change reduces coupling. Katello assumes there is RHSM support and this works towards that goal.

That said, I still think we should consider renaming it to `smart_proxy_content`. Today in the UI you see the Smart Proxy feature Pulpcore, but that's confusing. Especially if the version number of the Smart Proxy plugin doesn't match the Pulpcore version. I think we hacked around this now, but renaming it Content makes thing clearer. It's a shame we didn't do this straight away.